perception/detected_object_validation/src/object_lanelet_filter.cpp

```diff
@@ -0,0 +1,203 @@
+      Polygon2d polygon;
+      for (const auto & point : footprint.points) {
+        polygon.outer().emplace_back(point.x, point.y);
+      }
+      polygon.outer().push_back(polygon.outer().front());
+      if (isPolygonIntersectsLanelets(polygon, intersected_lanelets)) {
+        output_object_msg.objects.emplace_back(input_msg->objects.at(index));
+      }
+    } else {
+      output_object_msg.objects.emplace_back(input_msg->objects.at(index));
+    }
+    ++index;
+  }
+  object_pub_->publish(output_object_msg);
+}
+
+bool ObjectLaneletFilterNode::transformDetectedObjects(
+  const autoware_auto_perception_msgs::msg::DetectedObjects & input_msg,
+  const std::string & target_frame_id, const tf2_ros::Buffer & tf_buffer,
+  autoware_auto_perception_msgs::msg::DetectedObjects & output_msg)
+{
+  output_msg = input_msg;
+
+  /* transform to world coordinate */
+  if (input_msg.header.frame_id != target_frame_id) {
+    output_msg.header.frame_id = target_frame_id;
+    tf2::Transform tf_target2objects_world;
+    tf2::Transform tf_target2objects;
+    tf2::Transform tf_objects_world2objects;
+    {
+      const auto ros_target2objects_world =
+        getTransform(tf_buffer, input_msg.header.frame_id, target_frame_id, input_msg.header.stamp);
+      if (!ros_target2objects_world) {
+        return false;
+      }
+      tf2::fromMsg(*ros_target2objects_world, tf_target2objects_world);
+    }
+    for (auto & object : output_msg.objects) {
+      tf2::fromMsg(object.kinematics.pose_with_covariance.pose, tf_objects_world2objects);
+      tf_target2objects = tf_target2objects_world * tf_objects_world2objects;
+      tf2::toMsg(tf_target2objects, object.kinematics.pose_with_covariance.pose);
+    }
+  }
+  return true;
+}
+
+LinearRing2d ObjectLaneletFilterNode::getConvexHull(
+  const autoware_auto_perception_msgs::msg::DetectedObjects & detected_objects)
+{
+  MultiPoint2d candidate_points;
+  for (const auto & object : detected_objects.objects) {
+    for (const auto & p : object.shape.footprint.points) {
+      candidate_points.emplace_back(p.x, p.y);
+    }
+  }
+
+  LinearRing2d convex_hull;
+  boost::geometry::convex_hull(candidate_points, convex_hull);
+
+  return convex_hull;
+}
+
+lanelet::ConstLanelets ObjectLaneletFilterNode::getIntersectedLanelets(
+  const LinearRing2d & convex_hull, const lanelet::ConstLanelets & road_lanelets)
+{
+  lanelet::ConstLanelets intersected_lanelets;
+  for (const auto & road_lanelet : road_lanelets) {
+    if (boost::geometry::intersects(convex_hull, road_lanelet.polygon2d().basicPolygon())) {
+      intersected_lanelets.emplace_back(road_lanelet);
+    }
+  }
+  return intersected_lanelets;
+}
+bool ObjectLaneletFilterNode::isPolygonIntersectsLanelets(
+  const Polygon2d & polygon, const lanelet::ConstLanelets & intersected_lanelets)
+{
+  for (const auto & lanelet : intersected_lanelets) {
+    if (boost::geometry::intersects(polygon, lanelet.polygon2d().basicPolygon())) {
```

Comment:
If the obstacle polygon is within a lanelet polygon, is the obstacle removed?

Reply:
Right... It's not good at all. Let me think this again.